### PR TITLE
Record origin execution time as execution + output snapshotting time

### LINKED ISSUE #9117: Measure execution time more strictly
Currently we measure execution time for tasks and transforms by starting a timer as soon as we consider the work to be eligible for execution, and only stop it after the executed work's outputs have been snapshot again. This means that "execution time" includes the following:

* loading the previous state of the work from disk
* stale output cleanup
* snapshotting inputs (which might include downloading dependencies, or even executing artifact transforms)
* snapshotting outputs before execution
* actually executing the work
* snapshotting outputs after execution (but, interestingly, not the part where we store the resulting state on disk)

This can be misleading, especially when we use execution time to decide if it was worth loading something from cache, as we do many steps included in "execution time" also when we load things from cache, most important of which is the expensive snapshotting of inputs and outputs before execution.

Instead, we should measure only the actual execution and the snapshotting of the outputs afterwards as "execution time", as loading from cache only replaces those two steps.

### PR DESCRIPTION
**Pull request by ldaley**
Resolves #9117.

### CODE REVIEW

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Comment by ldaley** [comment]
@lptr @wolfs I'm at a bit of a loss for how to test this well. All advice appreciated.

**Comment by ldaley** [comment]
This PR currently shows my first attempt, then a revert of that, then a second simpler attempt.

**Comment by lptr** [comment]
Why include output snapshotting time?

**Comment by lptr** [comment]
I see, thanks. What I meant there was that we should calculate "execution time" used for the savings calculations by adding up the actual execution time and output snapshotting time. We should still measure them separately. I would do the adding up as late as possible, i.e. on the server side.

If this is complicated or not possible to do so with the current build scan model, I think it's okay to add them up in Gradle or in the build scan plugin for now, given that the current build operations only work for tasks anyway. It would be good to plot a course to where we can capture these measurements separately later on.

**Comment by ldaley** [comment]
> I see, thanks. What I meant there was that we should calculate "execution time" used for the savings calculations by adding up the actual execution time and output snapshotting time. We should still measure them separately. I would do the adding up as late as possible, i.e. on the server side.

Doing it on the server side adds significant complexity and reduce reliability, as currently we obtain this information from the artifact metadata, not the referenced scan. So, ultimately we need to record the number that we want at the build site. 

I'll rework this to combine discrete measuring of execution time and output snapshotting time as mentioned in my previous comment.

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Comment by ldaley** [comment]
@lptr ready for another look

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Review by lptr — CHANGES_REQUESTED** [review]
This looks good in general. I have one issue with the changes, and it's the introduction of `ExecuteWorkResult`. We already have quite a few confusing names around capturing the result of executed work that I'd love to clear up eventually:

- `ExecutionResult`
- `WorkResult`
- `Result`

This new name is a combination of the above without clarifying what its purpose is, increasing the confusion.

I see two ways to resolve this for the scope of the current PR:

1) just merge `Result` and `ExecuteWorkResult`: the latter replaced the former in most places anyway, and having the duration in a separate interface doesn't add anything,
2) rename `ExecuteWorkResult` to something like `TimedResult`.

I think we should go with 1).

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@lptr those types have been merged.

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by lptr** [comment]
> The situation with getDuration() being meaningless for some types seems awkward, but I didn't have any other ideas that wouldn't have resulted in major surgery.

One option to handle this better would be to return an `Optional<Duration>` instead, and use `Optional.empty()` for when the duration is meaningless. However, I don't think doing this is actually necessary.

I'd instead define `getDuration()` as the execution time of the work, no matter where or when it happened. That way it is pretty straightforward to implement it for all `Result` types. It also gives us a good place to document why the duration might be misleading.

Take a look at the changes here: https://gist.github.com/lptr/d75ac40b91a83db6736731a15a88c005

If we go in that direction, I'd probably also change `OriginMetadata.getExecutionTime()` to return `Duration` directly.

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@lptr patch applied.

**Review by britter — APPROVED** [review]
From what I can see this will not impact the Maven build cache implementation.

**Comment by ldaley** [comment]
@bot-gradle test RFM

File: subprojects/core/src/main/java/org/gradle/caching/internal/controller/impl/DefaultBuildCacheCommandFactory.java [hunk 1]
```diff
@@ -153,7 +155,7 @@ public BuildCacheKey getKey() {
 
         @Override
         public BuildCacheStoreCommand.Result store(OutputStream output) throws IOException {
-            final BuildCacheEntryPacker.PackResult packResult = packer.pack(entity, snapshots, output, originMetadataFactory.createWriter(entity, executionTime));
+            final BuildCacheEntryPacker.PackResult packResult = packer.pack(entity, snapshots, output, originMetadataFactory.createWriter(entity,  executionTime));
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/caching/internal/controller/impl/DefaultBuildCacheCommandFactory.java:158]: Nit: extra space.

**Review by lptr — APPROVED** [review]
LGTM once the tests are fixed.

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle help

**Comment by ldaley** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
I saw it. looking.

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by ldaley** [comment]
Thanks @blindpirate

**Pull request merged**